File: Stroll.History/Stroll.History.Mcp/Services/PerformanceMetrics.cs
using System.Collections.Concurrent;

namespace Stroll.History.Mcp.Services;

/// <summary>
/// Performance metrics collection for MCP service
/// 
/// Tracks and reports key performance indicators to demonstrate
/// improvements over the previous IPC implementation.
/// </summary>
public class PerformanceMetrics
{
    private readonly ConcurrentQueue<ToolCallMetric> _metrics = new();
    private readonly DateTime _startTime = DateTime.UtcNow;
    private readonly object _lock = new();
    
    private const int MaxMetrics = 10000; // Keep last 10k metrics in memory

    public void RecordToolCall(string toolName, double durationMs, bool success)
    {
        var metric = new ToolCallMetric
        {
            ToolName = toolName,
            Duration = durationMs,
            Success = success,
            Timestamp = DateTime.UtcNow
        };

        _metrics.Enqueue(metric);
        
        // Keep only recent metrics to prevent memory growth
        while (_metrics.Count > MaxMetrics)
        {
            _metrics.TryDequeue(out _);
        }
    }

    public PerformanceStats GetStats()
    {
        var metrics = _metrics.ToArray();
        
        if (metrics.Length == 0)
        {
            return new PerformanceStats
            {
                TotalRequests = 0,
                SuccessfulRequests = 0,
                FailedRequests = 0,
                AverageLatency = 0,
                P50Latency = 0,
                P95Latency = 0,
                P99Latency = 0,
                SuccessRate = 0,
                RequestsPerSecond = 0,
                ToolStats = new Dictionary<string, ToolStats>()
            };
        }

        var totalRequests = metrics.Length;
        var successfulRequests = metrics.Count(m => m.Success);
        var failedRequests = totalRequests - successfulRequests;

        // Calculate latency percentiles
        var sortedDurations = metrics.Select(m => m.Duration).OrderBy(d => d).ToArray();
        var averageLatency = sortedDurations.Average();
        var p50Latency = GetPercentile(sortedDurations, 0.5);
        var p95Latency = GetPercentile(sortedDurations, 0.95);
        var p99Latency = GetPercentile(sortedDurations, 0.99);

        // Calculate requests per second (last minute)
        var now = DateTime.UtcNow;
        var recentMetrics = metrics.Where(m => (now - m.Timestamp).TotalSeconds < 60).ToArray();
        var requestsPerSecond = recentMetrics.Length / 60.0;

        // Calculate per-tool statistics
        var toolStats = metrics
            .GroupBy(m => m.ToolName)
            .ToDictionary(
                g => g.Key,
                g => new ToolStats
                {
                    Count = g.Count(),
                    AverageLatency = g.Average(m => m.Duration),
                    SuccessRate = (double)g.Count(m => m.Success) / g.Count() * 100
                });

        return new PerformanceStats
        {
            TotalRequests = totalRequests,
            SuccessfulRequests = successfulRequests,
            FailedRequests = failedRequests,
            AverageLatency = averageLatency,
            P50Latency = p50Latency,
            P95Latency = p95Latency,
            P99Latency = p99Latency,
            SuccessRate = (double)successfulRequests / totalRequests * 100,
            RequestsPerSecond = requestsPerSecond,
            ToolStats = toolStats
        };
    }

    public string GetPerformanceReport()
    {
        var stats = GetStats();
        if (stats.TotalRequests == 0) return string.Empty;

        var uptime = DateTime.UtcNow - _startTime;
        
        var report = $@"
🚀 Stroll History MCP Performance Report
========================================

Uptime: {uptime.TotalSeconds:F1}s
Total Requests: {stats.TotalRequests}
Success Rate: {stats.SuccessRate:F2}%
Requests/sec: {stats.RequestsPerSecond:F1}

Latency Metrics:
- Average: {stats.AverageLatency:F2}ms
- P50: {stats.P50Latency:F2}ms
- P95: {stats.P95Latency:F2}ms  
- P99: {stats.P99Latency:F2}ms

Per-Tool Performance:";

        foreach (var (toolName, toolStat) in stats.ToolStats)
        {
            report += $"\n- {toolName}: {toolStat.Count} calls, {toolStat.AverageLatency:F2}ms avg, {toolStat.SuccessRate:F1}% success";
        }

        // Performance comparison with previous IPC
        report += $@"

Performance vs Previous IPC:
- Latency: {stats.AverageLatency:F2}ms (vs 200ms+ IPC) = {200 / Math.Max(stats.AverageLatency, 0.1):F1}x faster
- Success Rate: {stats.SuccessRate:F1}% (vs ~55% IPC) = {stats.SuccessRate / 55:F1}x more reliable  
- Throughput: {stats.RequestsPerSecond:F1} req/sec (vs ~5 req/sec IPC) = {stats.RequestsPerSecond / Math.Max(5, 0.1):F1}x higher";

        return report;
    }

    private static double GetPercentile(double[] sortedArray, double percentile)
    {
        if (sortedArray.Length == 0) return 0;
        if (sortedArray.Length == 1) return sortedArray[0];

        var index = percentile * (sortedArray.Length - 1);
        var lower = (int)Math.Floor(index);
        var upper = (int)Math.Ceiling(index);

        if (lower == upper) return sortedArray[lower];

        var weight = index - lower;
        return sortedArray[lower] * (1 - weight) + sortedArray[upper] * weight;
    }

    public void Reset()
    {
        lock (_lock)
        {
            while (_metrics.TryDequeue(out _)) { }
        }
    }
}

public record ToolCallMetric
{
    public required string ToolName { get; init; }
    public required double Duration { get; init; }
    public required bool Success { get; init; }
    public required DateTime Timestamp { get; init; }
}

public record PerformanceStats
{
    public required int TotalRequests { get; init; }
    public required int SuccessfulRequests { get; init; }
    public required int FailedRequests { get; init; }
    public required double AverageLatency { get; init; }
    public required double P50Latency { get; init; }
    public required double P95Latency { get; init; }
    public required double P99Latency { get; init; }
    public required double SuccessRate { get; init; }
    public required double RequestsPerSecond { get; init; }
    public required Dictionary<string, ToolStats> ToolStats { get; init; }
}

public record ToolStats
{
    public required int Count { get; init; }
    public required double AverageLatency { get; init; }
    public required double SuccessRate { get; init; }
}

File: Stroll.History/Stroll.History.Mcp/src/utils/performance-metrics.ts
/**
 * Performance Metrics Collection for Stroll History MCP Service
 * 
 * Tracks and reports performance metrics to demonstrate improvements over IPC.
 * Key metrics:
 * - Tool call latency (target: <5ms vs 200ms+ IPC)
 * - Throughput (target: 1000+ req/sec)
 * - Success rate (target: >99.5%)
 * - Memory usage patterns
 */

interface ToolCallMetric {
  name: string;
  duration: number;
  success: boolean;
  timestamp: number;
}

interface PerformanceStats {
  totalRequests: number;
  successfulRequests: number;
  failedRequests: number;
  averageLatency: number;
  p50Latency: number;
  p95Latency: number;
  p99Latency: number;
  successRate: number;
  requestsPerSecond: number;
  toolStats: Map<string, {
    count: number;
    averageLatency: number;
    successRate: number;
  }>;
}

export class PerformanceMetrics {
  private metrics: ToolCallMetric[] = [];
  private readonly maxMetrics = 10000; // Keep last 10k metrics
  private startTime: number;

  constructor() {
    this.startTime = Date.now();
  }

  recordToolCall(name: string, duration: number, success: boolean): void {
    const metric: ToolCallMetric = {
      name,
      duration,
      success,
      timestamp: Date.now(),
    };

    this.metrics.push(metric);

    // Keep only recent metrics to prevent memory growth
    if (this.metrics.length > this.maxMetrics) {
      this.metrics = this.metrics.slice(-this.maxMetrics);
    }
  }

  getStats(): PerformanceStats {
    if (this.metrics.length === 0) {
      return {
        totalRequests: 0,
        successfulRequests: 0,
        failedRequests: 0,
        averageLatency: 0,
        p50Latency: 0,
        p95Latency: 0,
        p99Latency: 0,
        successRate: 0,
        requestsPerSecond: 0,
        toolStats: new Map(),
      };
    }

    const totalRequests = this.metrics.length;
    const successfulRequests = this.metrics.filter(m => m.success).length;
    const failedRequests = totalRequests - successfulRequests;

    // Calculate latency percentiles
    const sortedDurations = this.metrics
      .map(m => m.duration)
      .sort((a, b) => a - b);

    const averageLatency = sortedDurations.reduce((a, b) => a + b, 0) / sortedDurations.length;
    const p50Latency = this.getPercentile(sortedDurations, 0.5);
    const p95Latency = this.getPercentile(sortedDurations, 0.95);
    const p99Latency = this.getPercentile(sortedDurations, 0.99);

    // Calculate requests per second
    const now = Date.now();
    const recentMetrics = this.metrics.filter(m => now - m.timestamp < 60000); // Last minute
    const requestsPerSecond = recentMetrics.length / 60;

    // Calculate per-tool statistics
    const toolStats = new Map<string, { count: number; averageLatency: number; successRate: number }>();
    const toolGroups = new Map<string, ToolCallMetric[]>();

    for (const metric of this.metrics) {
      if (!toolGroups.has(metric.name)) {
        toolGroups.set(metric.name, []);
      }
      toolGroups.get(metric.name)!.push(metric);
    }

    for (const [toolName, toolMetrics] of toolGroups) {
      const count = toolMetrics.length;
      const averageLatency = toolMetrics.reduce((sum, m) => sum + m.duration, 0) / count;
      const successCount = toolMetrics.filter(m => m.success).length;
      const successRate = (successCount / count) * 100;

      toolStats.set(toolName, { count, averageLatency, successRate });
    }

    return {
      totalRequests,
      successfulRequests,
      failedRequests,
      averageLatency,
      p50Latency,
      p95Latency,
      p99Latency,
      successRate: (successfulRequests / totalRequests) * 100,
      requestsPerSecond,
      toolStats,
    };
  }

  private getPercentile(sortedArray: number[], percentile: number): number {
    const index = percentile * (sortedArray.length - 1);
    const lower = Math.floor(index);
    const upper = Math.ceil(index);
    
    if (lower === upper) {
      return sortedArray[lower];
    }
    
    return sortedArray[lower] * (upper - index) + sortedArray[upper] * (index - lower);
  }

  getPerformanceReport(): string {
    const stats = this.getStats();
    const uptime = (Date.now() - this.startTime) / 1000;

    let report = `
🚀 Stroll History MCP Performance Report
========================================

Uptime: ${uptime.toFixed(1)}s
Total Requests: ${stats.totalRequests}
Success Rate: ${stats.successRate.toFixed(2)}%
Requests/sec: ${stats.requestsPerSecond.toFixed(1)}

Latency Metrics:
- Average: ${stats.averageLatency.toFixed(2)}ms
- P50: ${stats.p50Latency.toFixed(2)}ms  
- P95: ${stats.p95Latency.toFixed(2)}ms
- P99: ${stats.p99Latency.toFixed(2)}ms

Per-Tool Performance:
`;

    for (const [toolName, toolStat] of stats.toolStats) {
      report += `- ${toolName}: ${toolStat.count} calls, ${toolStat.averageLatency.toFixed(2)}ms avg, ${toolStat.successRate.toFixed(1)}% success\n`;
    }

    // Performance comparison with IPC
    report += `
Performance vs Previous IPC:
- Latency: ${stats.averageLatency.toFixed(2)}ms (vs 200ms+ IPC) = ${(200 / stats.averageLatency).toFixed(1)}x faster
- Success Rate: ${stats.successRate.toFixed(1)}% (vs ~55% IPC) = ${(stats.successRate / 55).toFixed(1)}x more reliable
- Throughput: ${stats.requestsPerSecond.toFixed(1)} req/sec (vs ~5 req/sec IPC) = ${(stats.requestsPerSecond / 5).toFixed(1)}x higher
`;

    return report;
  }

  exportMetrics(): ToolCallMetric[] {
    return [...this.metrics]; // Return copy
  }

  reset(): void {
    this.metrics = [];
    this.startTime = Date.now();
  }
}